### PR TITLE
Query Roles API

### PR DESCRIPTION
**Pull request by albertzaharovits**
This adds the Query Roles API:
```
POST /_security/_query/role
GET /_security/_query/role
```

This is similar to the currently existing:
 * [Query API key API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html)
 * [Query User API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-user.html)

Sample request:
```
POST /_security/_query/role
{
  "query": { 
    "bool": {
      "filter": [
        {
          "terms": {
            "applications.application": ["app-1", "app-2" ]
          }
        }
      ],
      "must_not": [
        {
          "match": {
            "description": {
              "query": "test match on role description (which is mapped as a text field)"
            }
          }
        }
      ]
    }
  },
  "sort": [ 
    "name"
  ],
  "search_after": [
    "role-name-1"
  ]
}
```

The query supports a subset of query types, including match_all, bool, term, terms, match, ids, prefix, wildcard, exists, range, and simple query string.

Currently, the supported fields are:
 * name
 * description
 * metadata
 * applications.application
 * applications.resources
 * applications.privileges

The query also supports pagination-related fields (`from`, `size`, `search_after`), analogous to the generic [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html).

The response format is similar to that of the [Query API key](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html) and [Query User](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-user.html) APIs. It contains a **list** of roles, in the sorted order (if specified). Unlike the [Get Roles API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-role.html), the role **name** is an attribute of the element in the list of roles (in the get-roles API case, the role name was the key in the response map, and the value was the rest of the role descriptor).
In addition, the element in the list of roles also contains the optional `_sort` field, eg (sample response):
```
{
  "total": 3,
  "count": 3,
  "roles": [
    {
      "name": "LYdz2",
      "cluster": [],
      "indices": [],
      "applications": [
        {
          "application": "ejYWvGQTF",
          "privileges": [
            "pRCfBMgOy",
            "zDhFtMQfc",
            "roudxado"
          ],
          "resources": [
            "nWHEpmgxy",
            "SOML/hMYrqx",
            "YIqP/*",
            "ueEomwsA"
          ]
        },
        {
          "application": "ampUW9",
          "privileges": [
            "jDvRtp"
          ],
          "resources": [
            "99"
          ]
        }
      ],
      "run_as": [],
      "metadata": {
        "nFKc": [
          1,
          0
        ],
        "PExF": [],
        "qlqY": -433239865,
        "IQXm": []
      },
      "transient_metadata": {
        "enabled": true
      },
      "description": "KoLlsEbq",
      "_sort": [
        "LYdz2"
      ]
    },
    {
      "name": "oaxW0",
      "cluster": [],
      "indices": [],
      "applications": [
        {
          "application": "*",
          "privileges": [
            "qZYb"
          ],
          "resources": [
            "tFrSULaKb"
          ]
        },
        {
          "application": "aLaEN9",
          "privileges": [
            "fCOc"
          ],
          "resources": [
            "gozqXtSgE",
            "UX/JgydeIM",
            "sjUp",
            "Ivdz/UAmuNrQAG"
          ]
        },
        {
          "application": "rbxyuKIMPAp",
          "privileges": [
            "lluqieFRu",
            "xKU",
            "gHlb"
          ],
          "resources": [
            "99"
          ]
        }
      ],
      "run_as": [],
      "metadata": {},
      "transient_metadata": {
        "enabled": true
      },
      "_sort": [
        "oaxW0"
      ]
    },
    {
      "name": "vWAV1",
      "cluster": [],
      "indices": [],
      "applications": [
        {
          "application": "*",
          "privileges": [
            "kWBWjCAc"
          ],
          "resources": [
            "hvEtV",
            "gZJ"
          ]
        },
        {
          "application": "avVUV9",
          "privileges": [
            "newZTa",
            "gQpxNm"
          ],
          "resources": [
            "99"
          ]
        }
      ],
      "run_as": [],
      "metadata": {},
      "transient_metadata": {
        "enabled": true
      },
      "_sort": [
        "vWAV1"
      ]
    }
  ]
}
```

### CODE REVIEW

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run elasticsearch-ci/rhel-8 / destructiveDistroTest.packages / packaging-tests-unix-sample

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FieldNameTranslators.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FieldNameTranslators.java:79]: It irks me a little that `application` is singular here but `resources` and `privileges` are plurals.
However, I think this is preferable because these are the names of these fields in the response.
WDYT @jfreden , should we make these all singulars?

Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FieldNameTranslators.java:79]: Yes, I agree, that's annoying. 

We can't change the current json serialization of a role response for backwards compatibility reasons and having it different for query roles doesn't make sense. 

I think it's more important that it's consistent with the response, so keeping `application` singular.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (jfreden) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:57]: What does this do? Can you add a comment explaining?

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (jfreden) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:37]: I think keeping this test focused and using the same meta key and value might make it easier to parse and leave the testing of the query logic to the `QueryRoleIT` suite.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (jfreden) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:163]: `migrationVersion` is always null so I think this can be removed.

**Review by jfreden — COMMENTED** [review]
👍 Really great job on this! I think adding documentation in this PR would be nice, but I'll leave that up to you.

I have nothing big, just a couple of smaller things. 

Don't forget to add a spec PR for the new API to: https://github.com/elastic/elasticsearch-specification

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/LicenseDLSFLSRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/LicenseDLSFLSRoleIT.java:40]: Yes, this is existing functionality. `transient_metadata.enabled` is false, when license does not allow DLS/FLS roles: https://github.com/elastic/elasticsearch/blob/d288dbf94ee8b327ead57861e012186104d692ae/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java#L455

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java:74]: We don't leak that information. The message is the same if the field exists or not: https://github.com/elastic/elasticsearch/blob/d288dbf94ee8b327ead57861e012186104d692ae/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/FieldNameTranslators.java#L309

Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java:74]: Do you still think a name change would be in order?

Comment (jfreden) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java:74]: Ah ok, then I think it's fine! Thanks for checking.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestQueryRoleAction.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestQueryRoleAction.java:37]: Good point! I just discussed it with @n1v0lg . I'll be removing the `ServerlessScope` annotation here, and worry about it and the `restrictPath` parameter later.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java:226]: I haven't thought about it before.
I added it, per your suggestion, although I'm not sure I prefer it, since it makes it harder to understand.
I hope maybe it avoids some unnecessary 503 to users...

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (jfreden) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:57]: Ah thanks for clarifying! For the things we're testing, does it make sense to cover that case? I'll leave it up to you but it feels like maybe it doesn't impact our implementation if one or two nodes are upgraded. Always good with some extra coverage though so I don't feel too strongly.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (albertzaharovits) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:163]: I'll do that. This is what's left of a past attempt to make the `waitForMigrationCompletion` cmd reusable, but couldn't find a place for it without changing the modules deps.

Comment (jfreden) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:163]: I see, that makes sense.

**Review by jfreden — APPROVED** [review]
I don't think there is anything blocking, so I'm going to approve (pending CI). Good job on this!

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java:421]: Yeah, OK, I'll follow your suggestion. I was at some point considering using it for descending sort.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (albertzaharovits) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:47]: >I'll leave it up to you but it feels like maybe it doesn't impact our implementation if one or two nodes are upgraded

Yeah, it's not crucial.
The issue that prompted me to change it was that the role, in a mixed cluster, was always double-created (which was odd).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleAction.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleAction.java]: Pushed [967294f](https://github.com/elastic/elasticsearch/pull/108733/commits/967294fbd116ebd349ed0902a2cf1cd89dbed12c)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleResponse.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleResponse.java:88]: I've added an assert in [04c03c5](https://github.com/elastic/elasticsearch/pull/108733/commits/04c03c57b8656e3154f38e899f887db9cb7ad9d3) (the role name is a doc _id , which cannot be null, so I'm not worried).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleAction.java
Comment (jfreden) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/QueryRoleAction.java]: Nice!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java]: Pushed [e6bb33c](https://github.com/elastic/elasticsearch/pull/108733/commits/e6bb33c99ff3f8d084dfb5fd273801ab9d4a5150)

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java]: Pushed [9f99438](https://github.com/elastic/elasticsearch/pull/108733/commits/9f994389a5fa8423356b7cd76ac5fc14f81430ac)

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (albertzaharovits) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java]: Pushed [12648cc](https://github.com/elastic/elasticsearch/pull/108733/commits/12648cc76fb1f386a0cb75e619ff395662a47be8)

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java
Comment (albertzaharovits) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRolesMetadataMigrationIT.java:35]: Pushed [d225966](https://github.com/elastic/elasticsearch/pull/108733/commits/d225966f37c771880f5aa5b446e21aa1741762d6)

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java:638]: Pushed [0d050fd](https://github.com/elastic/elasticsearch/pull/108733/commits/0d050fd36da862285e927a6fa51ca2e8f9079223)

File: x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java
Comment (albertzaharovits) [reply x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryRoleIT.java]: I renamed it to `getPrivilegeNameUsedForSorting` in [9f99438](https://github.com/elastic/elasticsearch/pull/108733/commits/9f994389a5fa8423356b7cd76ac5fc14f81430ac) .

**Comment by albertzaharovits** [comment]
I'll cover the docs and the rest specs from https://github.com/elastic/elasticsearch-specification in follow-ups soon.

**Pull request merged**
